server.cpp

```diff
@@ -40,38 +45,132 @@ size_t read_complete(const char* buff, const error_code& err, size_t bytes)
+        </table> \
+        </body> \
+        </html>";
+
+    const std::string index =  "HTTP/1.1 200 OK\r\nContent-Type: text/html; charset=utf-8\r\n\r\n" + table_html;
+
+    send_string(socket, index);
 }
 
-void write_response(tcp::socket& socket, const Request& request, const std::string& date)
+void write_file(tcp::socket& socket, const std::string& request_path_file, const std::string& path)
 {
-    std::string error_message;
+    int fd = open((path + "/" + request_path_file).c_str(), O_RDONLY);
+
+    if (fd == -1)
+    {
+        if (errno == ENOENT)
+            send_string(socket, "HTTP/1.1 404 File does not exist\r\nContent-Type: text/plain\r\n\r\n404 File does not exist.\n");
+        else if (errno == EACCES)
+            send_string(socket, "HTTP/1.1 403 File access not allowed\r\nContent-Type: text/plain\r\n\r\n403 File access not allowed.\n");
+        else
+            send_string(socket, "HTTP/1.1 500 File open error\r\nContent-Type: text/plain\r\n\r\n500 File open error." + std::string(strerror(errno)) + "\n");
+        return;
+    }
+    try
+    {
+        send_string(socket, "HTTP/1.1 200 OK\r\nContent-Type: application/octet-stream\r\nContent-Disposition: attachment\r\n\r\n");
+
+        char buff[1024] = {0};
+        size_t len;
+
+        while ((len = read(fd, buff, 1024)) > 0)
+            boost::asio::write(socket, boost::asio::buffer(buff, len));
+    }
+    catch (const std::exception& e)
+    {
+        close(fd);
+        throw e;
+    }
+    close(fd);
+}
 
+void write_response(tcp::socket& socket, const Request& request, const std::string& work_dir)
+{
     if (request.verb() != "GET")
-        error_message = "HTTP/1.1 405 Method not allowed\r\n";
+    {
+        send_string(socket, "HTTP/1.1 405 Method not allowed\r\nContent-Type: text/plain\r\n\r\n405 Method not allowed.\n");
+        return;
+    }
+
     if (request.version() != "HTTP/1.1" && request.version() != "HTTP/1.0")
-        error_message += "HTTP/1.1 505 HTTP Version Not Supported\r\n";
+    {
+        send_string(socket, "HTTP/1.1 505 HTTP Version Not Supported\r\nContent-Type: text/plain\r\n\r\n505 HTTP Version Not Supported.\n");
+        return;
+    }
 
-    error_code ignored_error;
+    const std::string path = work_dir.empty() ? "." : work_dir;
 
-    if (!error_message.empty())
+    if (request.path() == "/")
     {
-        boost::asio::write(socket, boost::asio::buffer(error_message), ignored_error);
+        DIR *dir = opendir(path.c_str());
+        if (dir == NULL)
+        {
+            send_string(socket, "HTTP/1.1 500 Failed to open directory\r\nContent-Type: text/plain\r\n\r\n500 Failed to open directory.\n");
+            return;
+        }
+        try
+        {
+            send_index(socket, dir, path);
+            closedir(dir);
+            return;
+        }
+        catch (const std::exception& e)
+        {
+            closedir(dir);
+            throw e;
```

Comment:
By throwing exception like this you introduce 'slicing'. Imagine a situation where `send_index` threw `std::runtime_error`. It get caught in your `catch`-clause because `std::exception` is a base class for `std::runtime_error`. So far so good. But when you throw it like you do - you re-throw `std::exception`, not the `std::runtime_error`. So it is impossible to catch the original exception later. Instead you should use the following form of `throw`:
```c++
throw;
```
Details: https://en.cppreference.com/w/cpp/language/throw

Take a look at the 'Notes' section there, it describes similar situation.

Also on 'slicing': https://en.wikipedia.org/wiki/Object_slicing